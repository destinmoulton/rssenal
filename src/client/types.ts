import { List, Map, OrderedMap, Set } from "immutable";
import { Notification } from "react-notification-system";
/** Redux Interfaces */
export interface IDispatch {
    <R>(
        asyncAction: (dispatch: IDispatch, getState: () => IRootStoreState) => R
    ): R;
    <R>(asyncAction: (dispatch: IDispatch) => R): R;
    // (neverAction: (dispatch: Dispatch, getState: () => GetState) => never): never;
    (action: object): void;
    // (action: Thunk): ; // thunks in this app must return a promise
}

export interface IGetState {
    (): IRootStoreState;
}

export interface IAuthAction {
    type: string;
    username?: string;
    password?: string;
    message?: string;
}

export type TEntryID = string;

export interface IEntry {
    _id: TEntryID;
    feed_id: string;
    guid: string;
    title: string;
    link: string;
    creator: string;
    content: string;
    content_snippet: string;
    publish_date: string;
    has_read: boolean;
    feedTitle?: string;
    timeAgo?: string;
}

export interface IEntriesAction {
    type: string;
    entries?: TEntries;
}

export type TEntries = OrderedMap<TEntryID, IEntry>;

/** Feeds Interfaces & Types */

export type TFeedID = string;

export interface IFeed {
    _id: TFeedID;
    title: string;
    folder_id: string;
    description: string;
    link: string;
    creation_date: string;
    last_updated: string;
}

export type TFeeds = OrderedMap<TFeedID, IFeed>;

export interface IFeedsAction {
    type: string;
    feeds?: TFeeds;
    unreadMap?: IFeedsUnreadMap;
}

export type TUnreadMapFeeds = Map<TFeedID, number>;
export type TUnreadMapFolders = Map<TFolderID, number>;
export interface IFeedsUnreadMap {
    entriesCounted: Set<TFeedID>;
    feeds: TUnreadMapFeeds;
    folders: TUnreadMapFolders;
}

/** Filter Interfaces */

export interface IFilter {
    id?: string;
    limit: string;
}
export interface IFilterAction {
    type: string;
    newFilter?: object;
    filterTitle?: string;
    filteredEntries?: TEntries;
}

/** Folders Interfaces & Types */

export type TFolderID = string;

export interface IFolder {
    _id: TFolderID;
    name: string;
    order: number;
}

export type TFolders = OrderedMap<TFolderID, IFolder>;

export interface IFolderAction {
    type: string;
    folders?: TFolders;
}

/** Messages Interfaces */
export interface IMessage {
    message: string;
    level: "error" | "warning" | "info" | "success";
    uid?: number;
    onRemove?: (message: IMessage) => void;
    position?: "br" | "tr" | "tl" | "tc" | "bl" | "bc";
}

export type TMessages = List<IMessage>;

export interface IMessageAction {
    type: string;
    messages: List<IMessage>;
    lastUID?: number;
}

/** Redux Reducer Interfaces */

export interface IReducerStateAuth {
    authenticationError: string;
    isAuthorized: boolean;
    isValidatingToken: boolean;
}

export interface IReducerStateEntries {
    entries: TEntries;
}

export interface IReducerStateFeeds {
    feeds: TFeeds;
    unreadMap: IFeedsUnreadMap;
    isAddingFeed: boolean;
    isUpdatingFeed: boolean;
}

export interface IReducerStateSettings {
    settings: TSettings;
}

export interface IReducerStateFilter {
    filter: IFilter;
    filterTitle: string;
    filteredEntries: TEntries;
}

export interface IReducerStateFolders {
    folders: TFolders;
    hasFolders: boolean;
    isDeletingFolder: boolean;
    isFetchingFolders: boolean;
    isSavingFolder: boolean;
}

export interface IReducerStateMessages {
    messages: TMessages;
    lastUID: 0;
}

export interface IRootStoreState {
    authStore: IReducerStateAuth;
    entriesStore: IReducerStateEntries;
    feedsStore: IReducerStateFeeds;
    filterStore: IReducerStateFilter;
    foldersStore: IReducerStateFolders;
    messagesStore: IReducerStateMessages;
    settingsStore: IReducerStateSettings;
}

/** Settings Interfaces */

export interface ISetting {
    name: string;
    type: string;
    refresh_entries_on_change: boolean;
    value: any;
}

export type TSettings = Map<string, ISetting>;

export interface ISettingsAction {
    type: string;
    settings?: TSettings;
}
