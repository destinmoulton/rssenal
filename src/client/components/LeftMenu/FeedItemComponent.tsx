import { Map } from "immutable";
import * as React from "react";

import { Icon } from "semantic-ui-react";

import { TFeedID, IFeed, IFeedsUnreadMap, IFilter } from "../../interfaces";

export interface IFeedItemMapState {
    filter: IFilter;
    unreadMapFeeds: Map<TFeedID, number>;
}

export interface IFeedItemMapDispatch {
    beginDeleteFeed: (feedId: TFeedID) => void;
    changeFilter: (newFilter: object) => void;
}

interface IFeedItemProps {
    editFeed: (feed: IFeed) => void;
    feed: IFeed;
}

type TAllProps = IFeedItemProps & IFeedItemMapState & IFeedItemMapDispatch;

interface IFeedItemState {
    isOptionsVisible: boolean;
}

class FeedItemComponent extends React.Component<TAllProps, IFeedItemState> {
    state = {
        isOptionsVisible: false
    };

    constructor(props: TAllProps) {
        super(props);

        this._handleClickDelete = this._handleClickDelete.bind(this);
        this._handleClickEdit = this._handleClickEdit.bind(this);
        this._handleClickFeed = this._handleClickFeed.bind(this);
        this._handleHideOptions = this._handleHideOptions.bind(this);
        this._handleShowOptions = this._handleShowOptions.bind(this);
    }

    _handleShowOptions() {
        this.setState({
            isOptionsVisible: true
        });
    }

    _handleHideOptions() {
        this.setState({
            isOptionsVisible: false
        });
    }

    _handleClickDelete() {
        const { beginDeleteFeed, feed } = this.props;
        const conf = confirm(
            `Are you sure you want to delete "${feed.title}"?`
        );
        if (conf) {
            beginDeleteFeed(feed._id);
        }
    }

    _handleClickEdit() {
        this.props.editFeed(this.props.feed);
    }

    _handleClickFeed() {
        const { changeFilter, feed } = this.props;

        changeFilter({
            limit: "feed",
            id: feed._id
        });
    }

    _buildOptions() {
        return (
            <div className="rss-folders-feeditem-options">
                <Icon
                    name="pencil"
                    color="green"
                    onClick={this._handleClickEdit}
                />
                <Icon
                    name="trash"
                    color="red"
                    onClick={this._handleClickDelete}
                />
            </div>
        );
    }

    render() {
        const { feed, filter, unreadMapFeeds } = this.props;
        const { isOptionsVisible } = this.state;

        let className = "rss-folders-feeditem";
        if (filter.limit === "feed" && filter.id === feed._id) {
            className = "rss-folders-feeditem rss-folders-feeditem-active";
        }

        let title = (
            <div
                className="rss-folders-feeditem-title"
                onClick={this._handleClickFeed}
            >
                {feed.title}
            </div>
        );

        let unreadEntriesCount = "";
        if (unreadMapFeeds.has(feed._id) && unreadMapFeeds.get(feed._id) > 0) {
            unreadEntriesCount = " [" + unreadMapFeeds.get(feed._id) + "]";
        }
        let unread = (
            <div
                className="rss-folders-feeditem-unread"
                onClick={this._handleClickFeed}
            >
                {unreadEntriesCount}
            </div>
        );

        let options: any = "";
        if (isOptionsVisible) {
            options = this._buildOptions();
        }

        return (
            <div
                className={className}
                onMouseEnter={this._handleShowOptions}
                onMouseLeave={this._handleHideOptions}
            >
                {title}
                {unread}
                {options}
            </div>
        );
    }
}

export default FeedItemComponent;
