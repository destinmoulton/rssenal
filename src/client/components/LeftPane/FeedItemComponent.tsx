import { Map } from "immutable";
import * as React from "react";

import { Icon } from "semantic-ui-react";

import { TFeedID, IFeed, IFilter, TUnreadMapFeeds } from "../../types";

export interface IFeedItemMapState {
    filter: IFilter;
    unreadMapFeeds: TUnreadMapFeeds;
}

export interface IFeedItemMapDispatch {
    deleteFeed: (feedId: TFeedID) => void;
    filterChange: (newFilter: object) => void;
    filterVisibleEntries: () => void;
}

interface IFeedItemProps {
    editFeed: (feed: IFeed) => void;
    feed: IFeed;
}

type TAllProps = IFeedItemProps & IFeedItemMapState & IFeedItemMapDispatch;

interface IFeedItemState {
    isOptionsVisible: boolean;
}

class FeedItemComponent extends React.Component<TAllProps, IFeedItemState> {
    state = {
        isOptionsVisible: false
    };

    _handleClickDelete = () => {
        const { deleteFeed, feed } = this.props;
        const conf = confirm(
            `Are you sure you want to delete "${feed.title}"?`
        );
        if (conf) {
            deleteFeed(feed._id);
        }
    };

    _handleClickEdit = () => {
        this.props.editFeed(this.props.feed);
    };

    _handleClickFeed = () => {
        const { filterChange, filterVisibleEntries, feed } = this.props;

        filterChange({
            limit: "feed",
            id: feed._id
        });

        filterVisibleEntries();
    };

    _handleHideOptions = () => {
        this.setState({
            isOptionsVisible: false
        });
    };

    _handleShowOptions = () => {
        this.setState({
            isOptionsVisible: true
        });
    };

    _buildOptions() {
        return (
            <div className="rss-folders-feeditem-options">
                <Icon
                    name="pencil"
                    color="green"
                    onClick={this._handleClickEdit}
                />
                <Icon
                    name="trash"
                    color="red"
                    onClick={this._handleClickDelete}
                />
            </div>
        );
    }

    render() {
        const { feed, filter, unreadMapFeeds } = this.props;
        const { isOptionsVisible } = this.state;

        let className = "rss-folders-feeditem";
        if (filter.limit === "feed" && filter.id === feed._id) {
            className = "rss-folders-feeditem rss-folders-feeditem-active";
        }

        let title = (
            <div
                className="rss-folders-feeditem-title"
                onClick={this._handleClickFeed}
            >
                {feed.title}
            </div>
        );

        let unreadEntriesCount = "";
        if (unreadMapFeeds.has(feed._id) && unreadMapFeeds.get(feed._id) > 0) {
            unreadEntriesCount = " [" + unreadMapFeeds.get(feed._id) + "]";
        }
        let unread = (
            <div
                className="rss-folders-feeditem-unread"
                onClick={this._handleClickFeed}
            >
                {unreadEntriesCount}
            </div>
        );

        let options: any = "";
        if (isOptionsVisible) {
            options = this._buildOptions();
        }

        return (
            <div
                className={className}
                onMouseEnter={this._handleShowOptions}
                onMouseLeave={this._handleHideOptions}
            >
                {title}
                {unread}
                {options}
            </div>
        );
    }
}

export default FeedItemComponent;
