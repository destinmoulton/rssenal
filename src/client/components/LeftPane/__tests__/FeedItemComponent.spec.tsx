import { Map } from "immutable";
import * as React from "react";

import { shallow } from "enzyme";
import * as Types from "../../../types";
import FeedItemComponent from "../FeedItemComponent";

import feed from "../../../../../test/data/feed";

const FILTER = {
    id: feed._id,
    limit: "feed"
};

const UNREAD_MAP: Types.TUnreadMapFeeds = Map([[feed._id, 10]]);
describe("<FeedItemComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <FeedItemComponent
                beginDeleteFeed={jest.fn()}
                filterChangeActive={jest.fn()}
                editFeed={jest.fn()}
                feed={feed}
                filter={FILTER}
                unreadMapFeeds={UNREAD_MAP}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
