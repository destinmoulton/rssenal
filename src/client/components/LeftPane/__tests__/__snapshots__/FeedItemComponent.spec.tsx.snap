// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedItemComponent /> renders and matches the snapshot 1`] = `
<div
  className="rss-folders-feeditem rss-folders-feeditem-active"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="rss-folders-feeditem-title"
    onClick={[Function]}
  >
    High Scalability
  </div>
  <div
    className="rss-folders-feeditem-unread"
    onClick={[Function]}
  >
     [10]
  </div>
</div>
`;
